### PR TITLE
Add broadcasting support (e.g. matrix-vector) for cos sim operator.

### LINKED ISSUE #3917: Add broadcasting support (e.g. Matrix-vector similarity) to cosine similarity operator.

### PR DESCRIPTION
**Pull request by xinghai-sun**
Resolve #3917

### CODE REVIEW

File: paddle/operators/cos_sim_op.h [hunk 1]
```diff
@@ -28,74 +28,114 @@ template <typename Place, typename T>
...
-    auto z = EigenMatrix<T>::From(*input_z);
-    auto x_norm = EigenMatrix<T>::From(*input_x_norm);
-    auto y_norm = EigenMatrix<T>::From(*input_y_norm);
-    auto dz = EigenMatrix<T>::From(*input_grad_z);
+    // convert Tensor to Eigen Tensor
+    int rows_x = in_x->dims()[0];
+    int rows_y = in_y->dims()[0];
+    int cols = framework::product(in_x->dims()) / rows_x;
+    auto x = EigenMatrix<T>::From(*in_x, framework::make_ddim({rows_x, cols}));
+    auto y = EigenMatrix<T>::From(*in_y, framework::make_ddim({rows_y, cols}));
```
Comment (pkuyym) [inline paddle/operators/cos_sim_op.h]: you can save ```framework::make_ddim({rows_x, cols})``` and ```framework::make_ddim({rows_y, cols})``` to  variables and reuse them in the followings.

File: paddle/operators/cos_sim_op.h [hunk 2]
```diff
@@ -28,74 +28,114 @@ template <typename Place, typename T>
...
+        auto dy = EigenMatrix<T>::From(*out_grad_y,
+                                       framework::make_ddim({rows_y, cols}));
+        auto grad = x / norm_prod_bcast - z_bcast * y / y_snorm_bcast;
+        dy.device(place) = dz_bcast * grad;
+      }
+    } else {
+      Eigen::DSizes<int, 2> bcast_row(rows_x, 1);
+      auto y_bcast =  y.broadcast(bcast_row);
+      auto y_snorm_bcast =
+          y_norm.square().eval().broadcast(bcast_row).eval().broadcast(bcast);
```
Comment (pkuyym) [inline paddle/operators/cos_sim_op.h]: Please merge the two ```broadcast``` operations to one like ```broadcast({rows, cols})```

File: paddle/operators/cos_sim_op.cc [hunk 3]
```diff
@@ -25,33 +25,51 @@ class CosSimOp : public framework::OperatorWithKernel {
...
 class CosSimOpMaker : public framework::OpProtoAndCheckerMaker {
  public:
   CosSimOpMaker(framework::OpProto *proto, framework::OpAttrChecker *op_checker)
       : OpProtoAndCheckerMaker(proto, op_checker) {
-    AddInput("X", "The first input of cos_sim op.");
-    AddInput("Y", "The second input of cos_sim op.");
+    AddInput("X", "The 1st input of cos_sim op.");
+    AddInput("Y", "The 2nd input of cos_sim op.");
     AddOutput("Out", "The output of cos_sim op.");
     AddOutput("XNorm", "Row norm of the first input.").AsIntermediate();
```
Comment (Canpio) [inline paddle/operators/cos_sim_op.cc]: What is `row norm`? Please add more instruction for this output.

File: paddle/operators/cos_sim_op.h [hunk 4]
```diff
@@ -28,74 +28,114 @@ template <typename Place, typename T>
...
-    auto x = EigenMatrix<T>::From(*input_x, new_dims);
-    auto y = EigenMatrix<T>::From(*input_y, new_dims);
-    auto z = EigenMatrix<T>::From(*output_z);
-    auto x_norm = EigenMatrix<T>::From(*output_x_norm);
-    auto y_norm = EigenMatrix<T>::From(*output_y_norm);
+    // convert Tensor to Eigen Tensor
+    int rows_x = in_x->dims()[0];
+    int rows_y = in_y->dims()[0];
+    int cols = framework::product(in_x->dims()) / rows_x;
+    auto x = EigenMatrix<T>::From(*in_x, framework::make_ddim({rows_x, cols}));
```
Comment (Canpio) [inline paddle/operators/cos_sim_op.h]: `EigenMatrix` has a member function `Reshape`, which converts a paddle tensor to EigenMatrix. 
```cpp
static typename EigenMatrix::ConstType Reshape(const Tensor& tensor, int num_col_dims);
```
The EigenMatrix's first dimension(column length) will be the product of tensor's first `num_col_dims` dimensions. In current case, it is 1.

See https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/eigen.h#L67

File: paddle/operators/cos_sim_op.cc [hunk 5]
```diff
@@ -62,32 +80,47 @@ class CosSimOpGrad : public framework::OperatorWithKernel {
...
-    PADDLE_ENFORCE_EQ(out_dims[1], 1, "1st dimension of Out@GRAD must be one.");
-
+    PADDLE_ENFORCE_EQ(ynorm_dims, framework::make_ddim({y_dims[0], 1}),
+                      "Shape of Input(YNorm) must be [Y.Dim(0), 1].");
+    auto out_dims = ctx.Input<Tensor>("Out")->dims();
+    PADDLE_ENFORCE_EQ(out_dims, framework::make_ddim({x_dims[0], 1}),
+                      "Shape of Input(Out) must be [X.Dim(0), 1].");
+    auto out_grad_dims =
+        ctx.Input<Tensor>(framework::GradVarName("Out"))->dims();
+    PADDLE_ENFORCE_EQ(out_grad_dims, framework::make_ddim({x_dims[0], 1}),
```
[the remark this answers is not in the document]
Comment (xinghai-sun) [reply paddle/operators/cos_sim_op.cc]: Done.

File: paddle/operators/cos_sim_op.h [hunk 6]
```diff
@@ -28,74 +28,114 @@ template <typename Place, typename T>
...
+    auto y = EigenMatrix<T>::From(*in_y, framework::make_ddim({rows_y, cols}));
+    auto z = EigenMatrix<T>::From(*out_z);
+    auto x_norm = EigenMatrix<T>::From(*out_x_norm);
+    auto y_norm = EigenMatrix<T>::From(*out_y_norm);
 
+    // compute
     auto place = context.GetEigenDevice<Place>();
-    auto xy = (x * y).sum(Eigen::array<int, 1>({1}));
     x_norm.device(place) = x.square().sum(Eigen::array<int, 1>({1})).sqrt();
```
[the remark this answers is not in the document]
Comment (xinghai-sun) [reply paddle/operators/cos_sim_op.h]: Done.

File: paddle/operators/cos_sim_op.h (hunk 2 above)
Comment (xinghai-sun) [reply paddle/operators/cos_sim_op.h]: Done.

File: paddle/operators/cos_sim_op.h (hunk 1 above)
Comment (xinghai-sun) [reply paddle/operators/cos_sim_op.h]: Done.

File: paddle/operators/cos_sim_op.h [hunk 7]
```diff
@@ -28,74 +28,114 @@ template <typename Place, typename T>
...
     y_norm.device(place) = y.square().sum(Eigen::array<int, 1>({1})).sqrt();
-    z.device(place) = xy / x_norm / y_norm;
+    if (rows_x == rows_y) {
+      auto xy = (x * y).sum(Eigen::array<int, 1>({1}));
+      z.device(place) = xy / x_norm / y_norm;
+    } else {
+      Eigen::DSizes<int, 2> bcast(rows_x, 1);
```
[the remark this answers is not in the document]
Comment (xinghai-sun) [reply paddle/operators/cos_sim_op.h]: to `bcast_rows`

File: paddle/operators/cos_sim_op.cc [hunk 8]
```diff
@@ -25,33 +25,51 @@ class CosSimOp : public framework::OperatorWithKernel {
...
+    // shape check
+    auto x_dims = ctx.Input<Tensor>("X")->dims();
+    auto y_dims = ctx.Input<Tensor>("Y")->dims();
+    PADDLE_ENFORCE_EQ(framework::arity(x_dims), framework::arity(y_dims),
+                      "Ranks of Input(X) and Input(Y) must be equal.");
+    PADDLE_ENFORCE_GE(framework::arity(x_dims), 2,
+                      "Rank of Input(X) must not be less than 2.");
+    PADDLE_ENFORCE_EQ(
+        framework::slice_ddim(x_dims, 1, framework::arity(x_dims)),
+        framework::slice_ddim(y_dims, 1, framework::arity(y_dims)),
```
[the remark this answers is not in the document]
Comment (Canpio) [reply paddle/operators/cos_sim_op.cc]: Got it!

File: paddle/operators/cos_sim_op.cc [hunk 9]
```diff
@@ -25,33 +25,51 @@ class CosSimOp : public framework::OperatorWithKernel {
...
-    ctx.Output<Tensor>("Out")->Resize({dims[0], 1});
-    ctx.Output<Tensor>("XNorm")->Resize({dims[0], 1});
-    ctx.Output<Tensor>("YNorm")->Resize({dims[0], 1});
+
+    // shape check
+    auto x_dims = ctx.Input<Tensor>("X")->dims();
+    auto y_dims = ctx.Input<Tensor>("Y")->dims();
+    PADDLE_ENFORCE_EQ(framework::arity(x_dims), framework::arity(y_dims),
+                      "Ranks of Input(X) and Input(Y) must be equal.");
+    PADDLE_ENFORCE_GE(framework::arity(x_dims), 2,
```
[the remark this answers is not in the document]
Comment (Canpio) [reply paddle/operators/cos_sim_op.cc]: Very reasonable! Thanks!

**Pull request merged**